src/services/kokkos/KokkosTime.cpp

```diff
@@ -0,0 +1,121 @@
+// Produced at the Lawrence Livermore National Laboratory.
+//
+// This file is part of Caliper.
+// Written by David Boehme, boehme3@llnl.gov.
+// LLNL-CODE-678900
+// All rights reserved.
+//
+// For details, see https://github.com/scalability-llnl/Caliper.
+// Please also see the LICENSE file for our additional BSD notice.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+//  * Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the disclaimer below.
+//  * Redistributions in binary form must reproduce the above copyright notice, this list of
+//    conditions and the disclaimer (as noted below) in the documentation and/or other materials
+//    provided with the distribution.
+//  * Neither the name of the LLNS/LLNL nor the names of its contributors may be used to endorse
+//    or promote products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
+// OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// LAWRENCE LIVERMORE NATIONAL SECURITY, LLC, THE U.S. DEPARTMENT OF ENERGY OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// KokkosLookup.cpp
+// Caliper kokkos variable lookup service
+
+#include "caliper/CaliperService.h"
+
+#include "caliper/Caliper.h"
+#include "../../caliper/MemoryPool.h"
+
+#include "caliper/common/Log.h"
+#include "caliper/common/RuntimeConfig.h"
+#include "caliper/Annotation.h"
+
+#include <iterator>
+#include <cstdint>
+
+#include "types.hpp"
+
+using namespace cali;
+
+namespace {
+
+    class KokkosTime {
+        static const ConfigSet::Entry s_configdata[];
+        cali::Annotation kernel_name_annot;
+        cali::Annotation kernel_type_annot;
+
+        KokkosTime(Caliper *c, Channel *chn) : kernel_name_annot("function"),
+                                               kernel_type_annot("kernel_type", CALI_ATTR_SKIP_EVENTS) {
+        }
+
+    public:
+
+        void pushRegion(const char *name, const char *type) {
+            kernel_type_annot.begin(type);
+            kernel_name_annot.begin(name);
+        }
+
+        void popRegion() {
+            kernel_type_annot.end();
+            kernel_name_annot.end();
```

Comment:
I'm not sure? I think the combination of `SKIP_EVENTS` and orderings to get exactly one trace record with both present looks like this, no? I always get confused about how to make one trace event happen with multiple Annotations

Reply:
You will get a begin and end record for `kernel_name`. However, if you end `kernel_type` before `kernel_end` you don't have the type anymore in the kernel end record. That's usually the more important one though. Also, it will mess up the context tree if you don't stack things correctly. So just do begin type, begin name, end name, end type.